Dockerfile.arm32v6

```diff
@@ -1,30 +1,49 @@
-FROM rust:1.40.0-slim-stretch AS builder
+# hadolint ignore=DL3007
+FROM balenalib/rpi-raspbian:latest AS builder
+
+ENV HOME /root/
 
 WORKDIR /gatekeeper
 
 COPY src            src
-COPY .cargo         .cargo
 COPY rust-toolchain .
 COPY Cargo.toml     .
 COPY Cargo.lock     .
 
-ENV CC_arm_unknown_linux_musleabihf arm-linux-gnueabihf-gcc
+RUN update-ca-certificates -f && \
+    curl -sSfL https://sh.rustup.rs > rustup.sh && \
+    sh rustup.sh -y \
+      --default-toolchain "$(cat rust-toolchain)-arm-unknown-linux-gnueabihf" \
+      --profile minimal \
+      --target arm-unknown-linux-gnueabihf && \
+    rm -f rustup.sh
+
+ENV PATH $HOME/.cargo/bin:$PATH
 
-# hadolint ignore=DL3015
 RUN apt-get update && \
-    apt-get install -y --install-recommends gcc-arm-linux-gnueabihf=* && \
+    apt-get install -y --no-install-recommends build-essential=* && \
     apt-get clean && \
-    rm -rf /var/lib/apt/lists/* && \
-    rustup target add arm-unknown-linux-musleabihf && \
-    cargo build --target=arm-unknown-linux-musleabihf --release && \
-    cp target/arm-unknown-linux-musleabihf/release/gatekeeperd /
+    rm -rf /var/lib/apt/lists/*
+
+ENV RUSTFLAGS "-Clinker=arm-linux-gnueabihf-gcc -Ccodegen_units=1"
```

Comment:
Is `codegen-units` correct?

```console
/gatekeeper$ rustc -C help | grep codegen
Available codegen options:
    -C            codegen-units=val -- divide crate into N units to optimize in parallel
```

Reply:
Why do you specify the option `codegen_units=1`?
The option is not specified in `Dockerfile.arm32v7`.

Reply:
Sorry, this is unnecessary.

Reply:
I'm not sure, but it seems that this specification is required to install Rust for armv6 from the source code.

Reply:
reproduction code:

(in gatekeeper directory on the Pi)

```sh
docker run -d -it -v "$PWD:/gatekeeper" --name=rpi-raspbian balenalib/rpi-raspbian:latest /bin/bash
docker exec -i rpi-raspbian /bin/bash -uex <<EOF
apt-get update
apt-get install -y build-essential
update-ca-certificates -f
curl -O 'https://static.rust-lang.org/dist/rust-1.40.0-arm-unknown-linux-gnueabihf.tar.gz'
tar xvzf rust-1.40.0-arm-unknown-linux-gnueabihf.tar.gz
./rust-1.40.0-arm-unknown-linux-gnueabihf/install.sh
cd /gatekeeper
cargo build --release # error
# RUSTFLAGS="-Ccodegen_units=1" cargo build --release
EOF
```

error log:

```
   Compiling syn v1.0.14
error: could not compile `unicode-xid`.

Caused by:
  process didn't exit successfully: `rustc --crate-name unicode_xid /root/.cargo/registry/src/github.com-1ecc6299db9ec823/unicode-xid-0.2.0/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C opt-level=3 --cfg 'feature="default"' -C metadata=71c55993a02b26e0 -C extra-filename=-71c55993a02b26e0 --out-dir /gatekeeper/target/release/deps -L dependency=/gatekeeper/target/release/deps --cap-lints allow` (signal: 11, SIGSEGV: invalid memory reference)
warning: build failed, waiting for other jobs to finish...
thread '<unnamed>' panicked at 'assertion failed: (*tail).value.is_none()', src/libstd/sync/mpsc/mpsc_queue.rs:97:17
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

malloc(): invalid next size (unsorted)
error: could not compile `syn`.

Caused by:
  process didn't exit successfully: `rustc --edition=2018 --crate-name build_script_build /root/.cargo/registry/src/github.com-1ecc6299db9ec823/syn-1.0.14/build.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type bin --emit=dep-info,link -C opt-level=3 --cfg 'feature="clone-impls"' --cfg 'feature="default"' --cfg 'feature="derive"' --cfg 'feature="extra-traits"' --cfg 'feature="parsing"' --cfg 'feature="printing"' --cfg 'feature="proc-macro"' --cfg 'feature="quote"' --cfg 'feature="visit"' -C metadata=37187c3e8bd59eb2 -C extra-filename=-37187c3e8bd59eb2 --out-dir /gatekeeper/target/release/build/syn-37187c3e8bd59eb2 -L dependency=/gatekeeper/target/release/deps --cap-lints allow` (signal: 6, SIGABRT: process abort signal)
warning: build failed, waiting for other jobs to finish...
error: failed to parse bitcode for LTO module: file doesn't start with bitcode header

thread '<unnamed>' panicked at 'src/librustc_codegen_ssa/back/write.rs:1488: worker thread panicked', src/librustc/util/bug.rs:37:26
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.40.0 (73528e339 2019-12-16) running on arm-unknown-linux-gnueabihf

note: compiler flags: -C opt-level=3 --crate-type bin

note: some of the compiler flags provided by cargo are hidden

error: aborting due to previous error

error: could not compile `libc`.
warning: build failed, waiting for other jobs to finish...
error: could not compile `proc-macro2`.

Caused by:
  process didn't exit successfully: `rustc --edition=2018 --crate-name build_script_build /root/.cargo/registry/src/github.com-1ecc6299db9ec823/proc-macro2-1.0.8/build.rs --error-format=json --json=diagnostic-rendered-ansi --crate-type bin --emit=dep-info,link -C opt-level=3 --cfg 'feature="default"' --cfg 'feature="proc-macro"' -C metadata=197115fbf6f734d5 -C extra-filename=-197115fbf6f734d5 --out-dir /gatekeeper/target/release/build/proc-macro2-197115fbf6f734d5 -L dependency=/gatekeeper/target/release/deps --cap-lints allow` (signal: 11, SIGSEGV: invalid memory reference)
```

Reply:
Maybe you're running out of memory?
`codegen-units` controlls code generation parallelism.

Reply:
Oh, I see.